pkg/auth/oauth2authenticator.go

```diff
@@ -136,15 +120,45 @@ func createVerifier(count int) []byte {
 	lut := []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")
 	verifier := make([]byte, count)
 
-	// TODO is this good enough?
 	for i := range verifier {
-		index := rand.Int() % len(lut)
+		index := randIndex(len(lut))
 		verifier[i] = lut[index]
 	}
 
 	return verifier
 }
 
+const maxRandIndexLimit = 256
+
+// randIndex provides a secure random number in the range [0, limit).
+//
+// Rejection sampling and modulo reduction are applied to avoid modulo bias and
+// ensure a uniform distribution.
+//
+// This function panics if the limit exceeds maxRandIndexLimit, or if a secure
+// random source cannot be read.
+//
+// Further reading:
+// https://research.kudelskisecurity.com/2020/07/28/the-definitive-guide-to-modulo-bias-and-how-to-avoid-it/
+// https://stackoverflow.com/a/10984975
+func randIndex(limit int) int {
+	if limit > maxRandIndexLimit {
+		panic(fmt.Sprintf("unsupported randIndex limit %d", limit))
+	}
+	rejectLimit := 256 - (256 % limit)
+	for {
+		var buf [1]byte
+		_, err := rand.Reader.Read(buf[:])
+		if err != nil {
+			panic(fmt.Sprintf("failed to read secure random bytes: %v", err))
```

Comment:
We could also return the error and wind that through `authenticateWithAuthorizationCode`, returning it from there. OTOH, if this fails, [your system is probably FUBAR](https://cs.opensource.google/go/go/+/refs/tags/go1.21.5:src/crypto/rand/rand.go;l=11).

I can go either way on this one.

Reply:
In general I would prefer to return errors also above in order to gracefully fail. sure we can try to gracefully fail on panics as well, but in my opinion, it is much clearer and easier with errors.

Reply:
I'm convinced, will update!